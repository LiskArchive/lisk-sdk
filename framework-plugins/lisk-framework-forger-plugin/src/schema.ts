/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const forgerInfoSchema = {
	$id: '/forger/info',
	type: 'object',
	properties: {
		totalProducedBlocks: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		totalReceivedFees: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		totalReceivedRewards: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		votesReceived: {
			type: 'array',
			items: {
				type: 'object',
				properties: {
					address: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					amount: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
				},
			},
			fieldNumber: 4,
			required: ['address', 'amount'],
		},
	},
	required: ['totalProducedBlocks', 'totalReceivedFees', 'totalReceivedRewards', 'votesReceived'],
};
