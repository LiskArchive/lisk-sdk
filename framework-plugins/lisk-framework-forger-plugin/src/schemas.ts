/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const forgerInfoSchema = {
	$id: '/forger/info',
	type: 'object',
	properties: {
		totalProducedBlocks: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		totalReceivedFees: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		totalReceivedRewards: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		stakeReceived: {
			type: 'array',
			fieldNumber: 4,
			items: {
				type: 'object',
				properties: {
					address: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					amount: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
				},
			},
			required: ['address', 'amount'],
		},
	},
	required: ['totalProducedBlocks', 'totalReceivedFees', 'totalReceivedRewards', 'stakeReceived'],
};

export const forgerSyncSchema = {
	$id: '/forger/sync',
	type: 'object',
	required: ['syncUptoHeight'],
	properties: {
		syncUptoHeight: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
	},
};

export const posVoteParamsSchema = {
	$id: '/forger/pos/stake',
	type: 'object',
	required: ['stakes'],
	properties: {
		stakes: {
			type: 'array',
			minItems: 1,
			maxItems: 20,
			items: {
				type: 'object',
				required: ['validatorAddress', 'amount'],
				properties: {
					validatorAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
						minLength: 20,
						maxLength: 20,
					},
					amount: {
						dataType: 'sint64',
						fieldNumber: 2,
					},
				},
			},
			fieldNumber: 1,
		},
	},
};
