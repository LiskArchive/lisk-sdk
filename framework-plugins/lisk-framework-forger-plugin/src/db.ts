/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import * as createDebug from 'debug';
import { codec, db as liskDB } from 'lisk-sdk';
import * as os from 'os';
import { join } from 'path';
import { ensureDir } from 'fs-extra';
import { DB_KEY_FORGER_INFO, DB_KEY_FORGER_SYNC_INFO } from './constants';
import { forgerInfoSchema, forgerSyncSchema } from './schemas';
import { ForgerInfo, ForgetSyncInfo } from './types';

const debug = createDebug('plugin:forger:db');

const { Database } = liskDB;
type KVStore = liskDB.Database;

export const getDBInstance = async (
	dataPath: string,
	dbName = 'lisk-framework-forger-plugin.db',
): Promise<KVStore> => {
	const dirPath = join(dataPath.replace('~', os.homedir()), 'plugins/data', dbName);
	await ensureDir(dirPath);

	return new Database(dirPath);
};

export const getForgerSyncInfo = async (db: KVStore): Promise<ForgetSyncInfo> => {
	try {
		const encodedSyncInfo = await db.get(DB_KEY_FORGER_SYNC_INFO);
		return codec.decode<ForgetSyncInfo>(forgerSyncSchema, encodedSyncInfo);
	} catch (error) {
		debug('Forger sync info does not exists');
		return {
			syncUptoHeight: 0,
		};
	}
};

export const setForgerSyncInfo = async (db: KVStore, blockHeight: number): Promise<void> => {
	const encodedSyncInfo = codec.encode(forgerSyncSchema, { syncUptoHeight: blockHeight });
	await db.set(DB_KEY_FORGER_SYNC_INFO, encodedSyncInfo);
};

export const setForgerInfo = async (
	db: KVStore,
	forgerAddress: string,
	forgerInfo: ForgerInfo,
): Promise<void> => {
	const encodedForgerInfo = codec.encode(forgerInfoSchema, forgerInfo);
	await db.set(
		Buffer.concat([DB_KEY_FORGER_INFO, Buffer.from(`:${forgerAddress}`, 'utf8')]),
		encodedForgerInfo,
	);
};

export const getForgerInfo = async (db: KVStore, forgerAddress: string): Promise<ForgerInfo> => {
	let forgerInfo;
	try {
		forgerInfo = await db.get(
			Buffer.concat([DB_KEY_FORGER_INFO, Buffer.from(`:${forgerAddress}`, 'utf8')]),
		);
	} catch (error) {
		debug(`Forger info does not exists for validator: ${forgerAddress}`);
		return {
			totalProducedBlocks: 0,
			totalReceivedFees: BigInt(0),
			totalReceivedRewards: BigInt(0),
			votesReceived: [],
		};
	}

	return codec.decode<ForgerInfo>(forgerInfoSchema, forgerInfo);
};
