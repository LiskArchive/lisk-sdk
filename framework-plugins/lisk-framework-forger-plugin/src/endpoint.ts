/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import {
	BasePlugin,
	BasePluginEndpoint,
	PluginEndpointContext,
	db as liskDB,
	cryptography,
} from 'lisk-sdk';
import { getForgerInfo } from './db';
import { Forger } from './types';

interface Staker {
	readonly address: string;
	readonly username: string;
	readonly totalStakeReceived: string;
	readonly stakers: {
		readonly address: string;
		readonly amount: string;
	}[];
}

interface ForgerInfo extends Forger {
	readonly username: string;
	readonly totalReceivedFees: string;
	readonly totalReceivedRewards: string;
	readonly totalProducedBlocks: number;
	readonly totalStakeReceived: string;
	readonly consecutiveMissedBlocks: number;
}

interface Validator {
	name: string;
	totalStakeReceived: string;
	selfStake: string;
	lastGeneratedHeight: number;
	isBanned: boolean;
	pomHeights: ReadonlyArray<number>;
	consecutiveMissedBlocks: number;
}

export class Endpoint extends BasePluginEndpoint {
	private _client!: BasePlugin['apiClient'];
	private _db!: liskDB.Database;

	public init(db: liskDB.Database, apiClient: BasePlugin['apiClient']) {
		this._db = db;
		this._client = apiClient;
	}

	public async getStakers(_context: PluginEndpointContext): Promise<Staker[]> {
		const forgersList = await this._client.invoke<Forger[]>('app_getForgingStatus');
		const forgerAccounts = [];
		for (const validator of forgersList) {
			const res = await this._client.invoke<Validator>('pos_getValidator', {
				address: validator.address,
			});
			forgerAccounts.push({
				...res,
				address: validator.address,
			});
		}

		const result: Staker[] = [];
		for (const account of forgerAccounts) {
			const forgerInfo = await getForgerInfo(
				this._db,
				Buffer.from(account.address, 'hex').toString('binary'),
			);

			result.push({
				address: account.address,
				username: account.name,
				totalStakeReceived: account.totalStakeReceived,
				stakers: forgerInfo.stakeReceived.map(stake => ({
					address: cryptography.address.getLisk32AddressFromAddress(stake.address),
					amount: stake.amount.toString(),
				})),
			});
		}

		return result;
	}

	public async getForgingInfo(_context: PluginEndpointContext): Promise<ForgerInfo[]> {
		const forgersList = await this._client.invoke<ReadonlyArray<Forger>>('app_getForgingStatus');
		const forgerAccounts = [];
		for (const validator of forgersList) {
			const res = await this._client.invoke<Validator>('pos_getValidator', {
				address: validator.address,
			});
			forgerAccounts.push({
				...res,
				address: validator.address,
			});
		}
		const data: ForgerInfo[] = [];
		for (const forgerAccount of forgerAccounts) {
			const forgerAddressBinary = Buffer.from(forgerAccount.address, 'hex').toString('binary');
			const forgerInfo = await getForgerInfo(this._db, forgerAddressBinary);
			const forger = forgersList.find(aForger => aForger.address === forgerAccount.address);

			if (forger) {
				data.push({
					...forger,
					username: forgerAccount.name,
					totalReceivedFees: forgerInfo.totalReceivedFees.toString(),
					totalReceivedRewards: forgerInfo.totalReceivedRewards.toString(),
					totalProducedBlocks: forgerInfo.totalProducedBlocks,
					totalStakeReceived: forgerAccount.totalStakeReceived,
					consecutiveMissedBlocks: forgerAccount.consecutiveMissedBlocks,
				});
			}
		}

		return data;
	}
}
