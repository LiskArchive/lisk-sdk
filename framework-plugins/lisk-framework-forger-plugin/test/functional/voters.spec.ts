/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { testing, PartialApplicationConfig } from 'lisk-sdk';
import { waitTill } from '../utils/application';
import { createVoteTransaction } from '../utils/transactions';
import { ForgerPlugin } from '../../src';

describe('forger:getStakers action', () => {
	let appEnv: testing.ApplicationEnv;
	let accountNonce = 0;
	let chainID: Buffer;

	beforeAll(async () => {
		const rootPath = '~/.lisk/forger-plugin';
		const config = {
			rootPath,
			label: 'stakers_functional',
		} as PartialApplicationConfig;

		appEnv = testing.createDefaultApplicationEnv({
			config,
			plugins: [new ForgerPlugin()],
		});
		await appEnv.startApplication();
		// The test application generates a dynamic genesis block so we need to get the networkID like this
		chainID = appEnv.application['_node'].chainID;
	});

	afterAll(async () => {
		// eslint-disable-next-line @typescript-eslint/no-empty-function
		jest.spyOn(process, 'exit').mockImplementation((() => {}) as never);
		await appEnv.stopApplication();
	});

	describe('action forger:getStakers', () => {
		it('should return valid format', async () => {
			// Arrange & Act
			const stakers = await appEnv.ipcClient.invoke('forger:getStakers');

			// Assert
			expect(stakers).toMatchSnapshot();
			expect(stakers).toBeInstanceOf(Array);
			expect(stakers).toHaveLength(103);
			expect(stakers[0]).toMatchObject(
				expect.objectContaining({
					address: expect.any(String),
					username: expect.any(String),
					totalStakeReceived: expect.any(String),
					stakers: expect.any(Array),
				}),
			);
		});

		it('should return valid stakers', async () => {
			// Arrange
			const initialVoters = await appEnv.ipcClient.invoke('forger:getStakers');
			const forgingDelegateAddress = (initialVoters[0] as any).address;
			const transaction = createVoteTransaction({
				amount: '10',
				recipientAddress: forgingDelegateAddress,
				fee: '0.3',
				nonce: accountNonce,
				chainID,
			});
			accountNonce += 1;

			await appEnv.ipcClient.invoke('txpool_postTransaction', {
				transaction: transaction.getBytes().toString('hex'),
			});
			await appEnv.waitNBlocks(1);
			// Wait a bit to give plugin a time to calculate forger info
			await waitTill(2000);

			// Act
			const stakers = await appEnv.ipcClient.invoke('forger:getStakers');
			const forgerInfo = (stakers as any).find(
				(forger: any) => forger.address === forgingDelegateAddress,
			);

			// Assert
			expect(forgerInfo.stakers[0]).toMatchObject(
				expect.objectContaining({
					address: transaction.senderAddress.toString('hex'),
					amount: '1000000000',
				}),
			);
		});
	});
});
