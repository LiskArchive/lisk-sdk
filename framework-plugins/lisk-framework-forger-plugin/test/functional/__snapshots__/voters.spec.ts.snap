// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forger:getStakers action action forger:getStakers should return valid format 1`] = `
Array [
  Object {
    "address": "9cabee3d27426676b852ce6b804cb2fdff7cd0b5",
    "totalStakeReceived": "0",
    "username": "validator_0",
    "stakers": Array [],
  },
  Object {
    "address": "463e7e879b7bdc6a97ec02a2a603aa1a46a04c80",
    "totalStakeReceived": "0",
    "username": "validator_1",
    "stakers": Array [],
  },
  Object {
    "address": "328d0f546695c5fa02105deb055cf2801d9b8ba1",
    "totalStakeReceived": "0",
    "username": "validator_2",
    "stakers": Array [],
  },
  Object {
    "address": "6174515fa66c91bff1128913edd4e0f1de37cee0",
    "totalStakeReceived": "0",
    "username": "validator_3",
    "stakers": Array [],
  },
  Object {
    "address": "9d0149b0962d44bfc08a9f64d5afceb6281d7fb5",
    "totalStakeReceived": "0",
    "username": "validator_4",
    "stakers": Array [],
  },
  Object {
    "address": "b56c55b9a70c8e2f07979b862374aed0e92a6dda",
    "totalStakeReceived": "0",
    "username": "validator_5",
    "stakers": Array [],
  },
  Object {
    "address": "b7580969dd56151f608931f126f793bbf45d8fa0",
    "totalStakeReceived": "0",
    "username": "validator_6",
    "stakers": Array [],
  },
  Object {
    "address": "58d907d26508603e838423daa2061c29c7a84950",
    "totalStakeReceived": "0",
    "username": "validator_7",
    "stakers": Array [],
  },
  Object {
    "address": "52f9cdcff0605241c78278690ae36eb0136a30ff",
    "totalStakeReceived": "0",
    "username": "validator_8",
    "stakers": Array [],
  },
  Object {
    "address": "b485becd88db1ab3d556d405204451ba00adaa7d",
    "totalStakeReceived": "0",
    "username": "validator_9",
    "stakers": Array [],
  },
  Object {
    "address": "4fd52f67f151fbbdda9dd92a714884a399830eca",
    "totalStakeReceived": "0",
    "username": "validator_10",
    "stakers": Array [],
  },
  Object {
    "address": "24c130eb6cc0d8f663a8f6d16ffc61f935a2e02e",
    "totalStakeReceived": "0",
    "username": "validator_11",
    "stakers": Array [],
  },
  Object {
    "address": "e39316cc020089ea7a5614bcf69a8931c10630a7",
    "totalStakeReceived": "0",
    "username": "validator_12",
    "stakers": Array [],
  },
  Object {
    "address": "6b9895c31bcdb2d9c929b9da7e389ed91de672a0",
    "totalStakeReceived": "0",
    "username": "validator_13",
    "stakers": Array [],
  },
  Object {
    "address": "1ac73bff74924ad9b74236c4962be27174ae87d0",
    "totalStakeReceived": "0",
    "username": "validator_14",
    "stakers": Array [],
  },
  Object {
    "address": "31fe789b43277e35ab410f2afcfb574280af2dd8",
    "totalStakeReceived": "0",
    "username": "validator_15",
    "stakers": Array [],
  },
  Object {
    "address": "a0620472cde03e77caece701ab7bc5928a5d367c",
    "totalStakeReceived": "0",
    "username": "validator_16",
    "stakers": Array [],
  },
  Object {
    "address": "f730cb929a1c45032387c345e10d2427bea55a5e",
    "totalStakeReceived": "0",
    "username": "validator_17",
    "stakers": Array [],
  },
  Object {
    "address": "a6f6a0543ae470c6b056021cb2ac153368eafeec",
    "totalStakeReceived": "0",
    "username": "validator_18",
    "stakers": Array [],
  },
  Object {
    "address": "79f30c1cbc1b9c4949c8b85acc24a7578e01558b",
    "totalStakeReceived": "0",
    "username": "validator_19",
    "stakers": Array [],
  },
  Object {
    "address": "cb579ee537b34926d47129a0b54c0e6d00ef3004",
    "totalStakeReceived": "0",
    "username": "validator_20",
    "stakers": Array [],
  },
  Object {
    "address": "436b40f58c0c27ed133f6001a019ff25561efad4",
    "totalStakeReceived": "0",
    "username": "validator_21",
    "stakers": Array [],
  },
  Object {
    "address": "4e874bcfb6f5896fe9e5dab3b26f59b2e2a9c09b",
    "totalStakeReceived": "0",
    "username": "validator_22",
    "stakers": Array [],
  },
  Object {
    "address": "d2c9a93755aed20c4d8f55c1e92b812d2c7d49d2",
    "totalStakeReceived": "0",
    "username": "validator_23",
    "stakers": Array [],
  },
  Object {
    "address": "3de95e18f18a54e2269bbf8f1a38ea70762c73fa",
    "totalStakeReceived": "0",
    "username": "validator_24",
    "stakers": Array [],
  },
  Object {
    "address": "df0e187bb3895806261c87cf66e1772566ee8e58",
    "totalStakeReceived": "0",
    "username": "validator_25",
    "stakers": Array [],
  },
  Object {
    "address": "be89f4e983dfb04e2b58a12eb9ed18149e108b07",
    "totalStakeReceived": "0",
    "username": "validator_26",
    "stakers": Array [],
  },
  Object {
    "address": "4b6126597881cb6ba1a45c1f6286769e7a094fb4",
    "totalStakeReceived": "0",
    "username": "validator_27",
    "stakers": Array [],
  },
  Object {
    "address": "70abf056bd92e8f77cfc551748fa54a4e3018d5f",
    "totalStakeReceived": "0",
    "username": "validator_28",
    "stakers": Array [],
  },
  Object {
    "address": "d0a0e45b950e3871d8783b973409042b4ab382d4",
    "totalStakeReceived": "0",
    "username": "validator_29",
    "stakers": Array [],
  },
  Object {
    "address": "3deeb0a7426a028b435b4ddd8d35ac85cf567237",
    "totalStakeReceived": "0",
    "username": "validator_30",
    "stakers": Array [],
  },
  Object {
    "address": "5fbd442a4647b079cda1229ecf6d8f44f361c8ca",
    "totalStakeReceived": "0",
    "username": "validator_31",
    "stakers": Array [],
  },
  Object {
    "address": "ad42f8e867d618171bf4982e64269442148f6e11",
    "totalStakeReceived": "0",
    "username": "validator_32",
    "stakers": Array [],
  },
  Object {
    "address": "03f6d90b7dbd0497dc3a52d1c27e23bb8c75897f",
    "totalStakeReceived": "0",
    "username": "validator_33",
    "stakers": Array [],
  },
  Object {
    "address": "9139c91f8a0aa1fb385770feaf299b99883aec2d",
    "totalStakeReceived": "0",
    "username": "validator_34",
    "stakers": Array [],
  },
  Object {
    "address": "61f396d2a4a13ab7a39ba791fac4b921b54a208e",
    "totalStakeReceived": "0",
    "username": "validator_35",
    "stakers": Array [],
  },
  Object {
    "address": "8722453383f781d5427a4ee211020e49bf34a2b9",
    "totalStakeReceived": "0",
    "username": "validator_36",
    "stakers": Array [],
  },
  Object {
    "address": "8074f0d02f748fc55448a4bf200f1dade8517059",
    "totalStakeReceived": "0",
    "username": "validator_37",
    "stakers": Array [],
  },
  Object {
    "address": "308a95d1d3f7bb556f48da4f4344566e59f6f1cb",
    "totalStakeReceived": "0",
    "username": "validator_38",
    "stakers": Array [],
  },
  Object {
    "address": "d5bd2050b74b309d54819ca17add173c6fca1e16",
    "totalStakeReceived": "0",
    "username": "validator_39",
    "stakers": Array [],
  },
  Object {
    "address": "8b1c221a030cf720736d9fb7d0499dd7276fc1b3",
    "totalStakeReceived": "0",
    "username": "validator_40",
    "stakers": Array [],
  },
  Object {
    "address": "6330fd8ae91df4a5d7fbc2390c182ec6676dc5a6",
    "totalStakeReceived": "0",
    "username": "validator_41",
    "stakers": Array [],
  },
  Object {
    "address": "94146c9889748c7b727eb3ac8c20e53c52effd32",
    "totalStakeReceived": "0",
    "username": "validator_42",
    "stakers": Array [],
  },
  Object {
    "address": "290abc4a2244bf0ecf5aa1ccee8ac8f60f8bce48",
    "totalStakeReceived": "0",
    "username": "validator_43",
    "stakers": Array [],
  },
  Object {
    "address": "abd2ed5ad35b3a0870aadae6dceacc988ba63895",
    "totalStakeReceived": "0",
    "username": "validator_44",
    "stakers": Array [],
  },
  Object {
    "address": "1c194c2be1cc53f663a93c64899cbaa34016f415",
    "totalStakeReceived": "0",
    "username": "validator_45",
    "stakers": Array [],
  },
  Object {
    "address": "5cd1d0ccf98f2bd5a4bfaa770d55f16498af0bcc",
    "totalStakeReceived": "0",
    "username": "validator_46",
    "stakers": Array [],
  },
  Object {
    "address": "657f610728eef97d55e50212871f0993bb7cc700",
    "totalStakeReceived": "0",
    "username": "validator_47",
    "stakers": Array [],
  },
  Object {
    "address": "2159f75e5440c36431aedbc7dc29a65a327778b8",
    "totalStakeReceived": "0",
    "username": "validator_48",
    "stakers": Array [],
  },
  Object {
    "address": "0f33a5033b750e6c4dca47e38ba020e912df143e",
    "totalStakeReceived": "0",
    "username": "validator_49",
    "stakers": Array [],
  },
  Object {
    "address": "9b42e4264020f3c3dcaaed806578ccd469205060",
    "totalStakeReceived": "0",
    "username": "validator_50",
    "stakers": Array [],
  },
  Object {
    "address": "aebd99f07218109162a905d0e0c91e58bedc83c5",
    "totalStakeReceived": "0",
    "username": "validator_51",
    "stakers": Array [],
  },
  Object {
    "address": "82cbc7b39d35af358f9e2513af13b2f77b647a00",
    "totalStakeReceived": "0",
    "username": "validator_52",
    "stakers": Array [],
  },
  Object {
    "address": "0d2c377e936b68c70066613b10c0fdad537f90da",
    "totalStakeReceived": "0",
    "username": "validator_53",
    "stakers": Array [],
  },
  Object {
    "address": "89b144ecfdd5ea352083bf624d3cf842ec06a5e3",
    "totalStakeReceived": "0",
    "username": "validator_54",
    "stakers": Array [],
  },
  Object {
    "address": "2cf52c08cc76091d884e800c1c697b13f69907d4",
    "totalStakeReceived": "0",
    "username": "validator_55",
    "stakers": Array [],
  },
  Object {
    "address": "e2950a9f07b44e724df2129360cc140293c08308",
    "totalStakeReceived": "0",
    "username": "validator_56",
    "stakers": Array [],
  },
  Object {
    "address": "65f927187bf96aac5d968fcc9351e5492b5f9356",
    "totalStakeReceived": "0",
    "username": "validator_57",
    "stakers": Array [],
  },
  Object {
    "address": "d06fe6d3e5f7facb5855eca839422fe3824a5d6e",
    "totalStakeReceived": "0",
    "username": "validator_58",
    "stakers": Array [],
  },
  Object {
    "address": "6e12e4498ae69fb07ff2d8aab036a911229d6c62",
    "totalStakeReceived": "0",
    "username": "validator_59",
    "stakers": Array [],
  },
  Object {
    "address": "dcb5bf35b6d521195e613c42483f520139e2331d",
    "totalStakeReceived": "0",
    "username": "validator_60",
    "stakers": Array [],
  },
  Object {
    "address": "8506f3c10f75044946f1a23a7caf578253649471",
    "totalStakeReceived": "0",
    "username": "validator_61",
    "stakers": Array [],
  },
  Object {
    "address": "d3c8064d011ef853e3be506b95a045f41f78e72a",
    "totalStakeReceived": "0",
    "username": "validator_62",
    "stakers": Array [],
  },
  Object {
    "address": "c98554123062ac5795a3ee905b081e863db5a818",
    "totalStakeReceived": "0",
    "username": "validator_63",
    "stakers": Array [],
  },
  Object {
    "address": "a0bc50b27e7ac39060ed015a55f2f4508c84f0c2",
    "totalStakeReceived": "0",
    "username": "validator_64",
    "stakers": Array [],
  },
  Object {
    "address": "d5c4e380b1ec2f7f2068cfba9a90cb3ae7816110",
    "totalStakeReceived": "0",
    "username": "validator_65",
    "stakers": Array [],
  },
  Object {
    "address": "7d2c6781d873ed2ba7a87f46f735f5e15a41a6f1",
    "totalStakeReceived": "0",
    "username": "validator_66",
    "stakers": Array [],
  },
  Object {
    "address": "3b3e137b1bec6f20c9a8b2ad4f5784661fb0fa79",
    "totalStakeReceived": "0",
    "username": "validator_67",
    "stakers": Array [],
  },
  Object {
    "address": "6ffcd8ad547d8a549a31b25236e322c781a52d85",
    "totalStakeReceived": "0",
    "username": "validator_68",
    "stakers": Array [],
  },
  Object {
    "address": "bd175729d4177259c71cf13fd4ecfb5d01542706",
    "totalStakeReceived": "0",
    "username": "validator_69",
    "stakers": Array [],
  },
  Object {
    "address": "5853a3f24990deecced49d6bc15990102ec0c33a",
    "totalStakeReceived": "0",
    "username": "validator_70",
    "stakers": Array [],
  },
  Object {
    "address": "e9355152c117c9e1fad8be86e9abea961cef4a36",
    "totalStakeReceived": "0",
    "username": "validator_71",
    "stakers": Array [],
  },
  Object {
    "address": "4f4422eb61c45edb4d76f10cd871c9f983f2ebaa",
    "totalStakeReceived": "0",
    "username": "validator_72",
    "stakers": Array [],
  },
  Object {
    "address": "0903f4c5cb599a7928aef27e314e98291d1e3888",
    "totalStakeReceived": "0",
    "username": "validator_73",
    "stakers": Array [],
  },
  Object {
    "address": "31204ad5b95dd922c2899aa5bf8e7ee5b7546af3",
    "totalStakeReceived": "0",
    "username": "validator_74",
    "stakers": Array [],
  },
  Object {
    "address": "8459b8870fcefff59f172d716b7bfe9fcc28d408",
    "totalStakeReceived": "0",
    "username": "validator_75",
    "stakers": Array [],
  },
  Object {
    "address": "246fba5c519576d93c5fac899c44b29b72f526ae",
    "totalStakeReceived": "0",
    "username": "validator_76",
    "stakers": Array [],
  },
  Object {
    "address": "d8e611bafd70a549f035cf61ab0d6ed9e7f25c4e",
    "totalStakeReceived": "0",
    "username": "validator_77",
    "stakers": Array [],
  },
  Object {
    "address": "7d60db187337cbd881140d69d84c9246eda8382e",
    "totalStakeReceived": "0",
    "username": "validator_78",
    "stakers": Array [],
  },
  Object {
    "address": "b11c5811ea074a30142d824b6e8cfd3df14b2688",
    "totalStakeReceived": "0",
    "username": "validator_79",
    "stakers": Array [],
  },
  Object {
    "address": "8eceffd5a41e678b6467c9bc80ce35d2e8543d98",
    "totalStakeReceived": "0",
    "username": "validator_80",
    "stakers": Array [],
  },
  Object {
    "address": "ca5f6d76eab6e4f5aacee2864c79034d7111b986",
    "totalStakeReceived": "0",
    "username": "validator_81",
    "stakers": Array [],
  },
  Object {
    "address": "0bc3bec2fdb565996fd316e368e66e5d8e830808",
    "totalStakeReceived": "0",
    "username": "validator_82",
    "stakers": Array [],
  },
  Object {
    "address": "ab0041a7d3f7b2c290b5b834d46bdc7b7eb85815",
    "totalStakeReceived": "0",
    "username": "validator_83",
    "stakers": Array [],
  },
  Object {
    "address": "c697b620c7c4015e32dd7bdd7d0430b33404e107",
    "totalStakeReceived": "0",
    "username": "validator_84",
    "stakers": Array [],
  },
  Object {
    "address": "c3ab2ac23512d9bf62b02775e22cf80df814eb1b",
    "totalStakeReceived": "0",
    "username": "validator_85",
    "stakers": Array [],
  },
  Object {
    "address": "ca309a5f4bbf11ca86592febb6d2ccc78309f69e",
    "totalStakeReceived": "0",
    "username": "validator_86",
    "stakers": Array [],
  },
  Object {
    "address": "27843a60a1e044c1e6e3cf119fdf64eb2b3e0d94",
    "totalStakeReceived": "0",
    "username": "validator_87",
    "stakers": Array [],
  },
  Object {
    "address": "8ac800124d5b16afd57b5cf7245edfcd5885ea3b",
    "totalStakeReceived": "0",
    "username": "validator_88",
    "stakers": Array [],
  },
  Object {
    "address": "a28d5e34007fd8fe6d7903044eb23a60fdad3c00",
    "totalStakeReceived": "0",
    "username": "validator_89",
    "stakers": Array [],
  },
  Object {
    "address": "d5e1f52cbe4a11a3730b98f52109b57602a9c4a1",
    "totalStakeReceived": "0",
    "username": "validator_90",
    "stakers": Array [],
  },
  Object {
    "address": "3c80e7d9964a1c83a6dd5dc64e105e0e634bd58a",
    "totalStakeReceived": "0",
    "username": "validator_91",
    "stakers": Array [],
  },
  Object {
    "address": "fa526a1611ccc66dec815cb963174118074b736e",
    "totalStakeReceived": "0",
    "username": "validator_92",
    "stakers": Array [],
  },
  Object {
    "address": "3b96d8565569421f43684b2c4eaa0639cbb5e011",
    "totalStakeReceived": "0",
    "username": "validator_93",
    "stakers": Array [],
  },
  Object {
    "address": "acfbdbaeb93d587170c7cd9c0b5ffdeb7ff9daec",
    "totalStakeReceived": "0",
    "username": "validator_94",
    "stakers": Array [],
  },
  Object {
    "address": "ffce8ce225c5d80098f50e877125b655aef6d101",
    "totalStakeReceived": "0",
    "username": "validator_95",
    "stakers": Array [],
  },
  Object {
    "address": "a9c66694dd65b2fdf40cdf45a0c308cbd38004fc",
    "totalStakeReceived": "0",
    "username": "validator_96",
    "stakers": Array [],
  },
  Object {
    "address": "0ada6a2f6c8f891769366fc9aa6fd9f1facb36cf",
    "totalStakeReceived": "0",
    "username": "validator_97",
    "stakers": Array [],
  },
  Object {
    "address": "b76a0f1819c4be0a1482567ca9b9fbed3eda444c",
    "totalStakeReceived": "0",
    "username": "validator_98",
    "stakers": Array [],
  },
  Object {
    "address": "b543e2e592200beb38235f6e48f8abe1d87ad872",
    "totalStakeReceived": "0",
    "username": "validator_99",
    "stakers": Array [],
  },
  Object {
    "address": "4fd8cc4e27a3489b57ed986efe3d327d3de40d92",
    "totalStakeReceived": "0",
    "username": "validator_100",
    "stakers": Array [],
  },
  Object {
    "address": "936f3a0f4d776b6a7722ed126e8ff17b44d7e7b8",
    "totalStakeReceived": "0",
    "username": "validator_101",
    "stakers": Array [],
  },
  Object {
    "address": "5ade564399e670bd1d429583059067f3a6ca2b7f",
    "totalStakeReceived": "0",
    "username": "validator_102",
    "stakers": Array [],
  },
]
`;
